habitat_baselines/rl/ppo/ppo_trainer.py

```diff
@@ -188,283 +402,411 @@ def _collect_rollout_step(
         for k, v_k in self._extract_scalars_from_infos(infos).items():
             v = torch.tensor(
-                v_k, dtype=torch.float, device=current_episode_reward.device
+                v_k,
+                dtype=torch.float,
+                device=self.current_episode_reward.device,
             ).unsqueeze(1)
-            if k not in running_episode_stats:
-                running_episode_stats[k] = torch.zeros_like(
-                    running_episode_stats["count"]
+            if k not in self.running_episode_stats:
+                self.running_episode_stats[k] = torch.zeros_like(
+                    self.running_episode_stats["count"]
                 )
 
-            running_episode_stats[k] += (1 - masks) * v  # type: ignore
+            self.running_episode_stats[k][env_slice] += v.where(done_masks, v.new_zeros(()))  # type: ignore
 
-        current_episode_reward *= masks
+        self.current_episode_reward[env_slice].masked_fill_(done_masks, 0.0)
 
         if self._static_encoder:
             with torch.no_grad():
                 batch["visual_features"] = self._encoder(batch)
 
-        rollouts.insert(
-            batch,
-            recurrent_hidden_states,
-            actions,
-            actions_log_probs,
-            values,
-            rewards,
-            masks,
+        self.rollouts.insert(
+            next_observations=batch,
+            rewards=rewards,
+            next_masks=not_done_masks,
+            buffer_index=buffer_index,
         )
 
-        pth_time += time.time() - t_update_stats
+        self.rollouts.advance_rollout(buffer_index)
 
-        return pth_time, env_time, self.envs.num_envs
+        self.pth_time += time.time() - t_update_stats
+
+        return env_slice.stop - env_slice.start
+
+    @profiling_wrapper.RangeContext("_collect_rollout_step")
+    def _collect_rollout_step(self):
+        self._compute_actions_and_step_envs()
+        return self._collect_environment_result()
 
     @profiling_wrapper.RangeContext("_update_agent")
-    def _update_agent(self, ppo_cfg, rollouts):
+    def _update_agent(self):
+        ppo_cfg = self.config.RL.PPO
         t_update_model = time.time()
         with torch.no_grad():
-            last_observation = {
-                k: v[rollouts.step] for k, v in rollouts.observations.items()
-            }
+            step_batch = self.rollouts.buffers[self.rollouts.step]
+
             next_value = self.actor_critic.get_value(
-                last_observation,
-                rollouts.recurrent_hidden_states[rollouts.step],
-                rollouts.prev_actions[rollouts.step],
-                rollouts.masks[rollouts.step],
-            ).detach()
+                step_batch["observations"],
+                step_batch["recurrent_hidden_states"],
+                step_batch["prev_actions"],
+                step_batch["masks"],
+            )
```

Comment:
Curious, why no need in `.detach` anymore?

Reply:
It was never necessary.  This is all run under `with torch.no_grad():` so there is not autograd graph to detach from.